File: backend/routes/notifications.js
const express = require('express');
const router = express.Router();

// In-memory token store (use Redis/DB in production)
const deviceTokens = new Map(); // userId → [tokens]

/**
 * POST /api/notifications/register-token
 * Saves FCM device token for a user
 */
router.post('/register-token', (req, res) => {
  const { userId, token, platform } = req.body;
  if (!userId || !token) {
    return res.status(400).json({ success: false, error: 'userId and token required' });
  }

  const existing = deviceTokens.get(userId) || [];
  if (!existing.includes(token)) {
    existing.push(token);
    deviceTokens.set(userId, existing);
  }

  console.log(`[FCM] Token registered for user ${userId} (${platform || 'web'}) — total: ${existing.length}`);
  res.json({ success: true, message: 'Token registered' });
});

/**
 * Internal helper — called from crisis route when SOS is triggered
 * Sends push notification to all hospital staff tokens
 */
async function sendPushToHospital(hospitalId, payload) {
  try {
    const admin = require('../services/firebaseAdmin');
    if (!admin.isConfigured()) return;

    const hospitalTokens = getHospitalTokens(hospitalId);
    if (!hospitalTokens.length) return;

    const message = {
      notification: {
        title: payload.title || '🚨 Emergency SOS Alert',
        body:  payload.body  || 'New patient incoming — action required',
      },
      data: {
        incidentId:   payload.incidentId || '',
        patientName:  payload.patientName || '',
        severity:     payload.severity || 'unknown',
        url:          '/hospital/portal',
      },
      tokens: hospitalTokens,
    };

    const response = await admin.messaging().sendEachForMulticast(message);
    console.log(`[FCM] Push sent: ${response.successCount} success, ${response.failureCount} failed`);
  } catch (err) {
    console.warn('[FCM] Push notification failed (non-fatal):', err.message);
  }
}

// Mock hospital → staff token mapping (replace with DB in production)
function getHospitalTokens(hospitalId) {
  // Return all registered tokens for now (in production: filter by hospital)
  const all = [];
  deviceTokens.forEach(tokens => all.push(...tokens));
  return all;
}

router.sendPushToHospital = sendPushToHospital;

module.exports = router;

File: backend/routes/notifications.py
"""
routes/notifications.py — FCM device token registration and push notifications
Mirrors backend/routes/notifications.js
"""
from typing import Optional, List, Dict
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from services.firebase_admin import send_push_multicast, is_configured

router = APIRouter(prefix="/api/notifications", tags=["Notifications"])

device_tokens: Dict[str, List[str]] = {}


class RegisterTokenRequest(BaseModel):
    userId: str
    token: str
    platform: Optional[str] = "web"


@router.post("/register-token")
async def register_token(payload: RegisterTokenRequest):
    if not payload.userId or not payload.token:
        raise HTTPException(status_code=400, detail="userId and token required")

    tokens = device_tokens.setdefault(payload.userId, [])
    if payload.token not in tokens:
        tokens.append(payload.token)

    print(f"[FCM] Token registered for user {payload.userId} ({payload.platform}) — total: {len(tokens)}")
    return {"success": True, "message": "Token registered"}


def get_all_tokens() -> List[str]:
    all_tokens = []
    for tokens in device_tokens.values():
        all_tokens.extend(tokens)
    return list(set(all_tokens))


async def send_push_to_hospital(hospital_id: str, payload: Dict[str, Any]):
    if not is_configured():
        return

    tokens = get_all_tokens()
    if not tokens:
        return

    notification = {
        "title": payload.get("title", "🚨 Emergency SOS Alert"),
        "body": payload.get("body", "New patient incoming — action required"),
    }
    data = {
        "incidentId": str(payload.get("incidentId", "")),
        "patientName": str(payload.get("patientName", "")),
        "severity": str(payload.get("severity", "unknown")),
        "url": "/hospital/portal",
    }
    try:
        res = send_push_multicast(tokens, notification, data)
        print(f"[FCM] Push sent: {res}")
    except Exception as e:
        print(f"[FCM] Push notification failed: {e}")
